### PR TITLE
chore(controller): remove $PORT logic from fleet

### LINKED ISSUE #1156: controller exposes all container port bindings instead of first
tested on deis revision: df1c0edd7f9f44999da5d11d5f2c674bf0b17a78  Tue Jun 10 14:00:57 2014 -0700

CoreOS 343.0.0

deis executes the following code:
 core@core1 ~ $ docker inspect -f '{{range $i, $e := .HostConfig.PortBindings }}{{$p := index $e 0}}{{$p.HostPort}}{{end}}' wp_v2.cmd.1
4915349154

the result is a concatenation of the 2 ports which obiously doesn't work.

### PR DESCRIPTION
**Pull request by bacongobbler**
For buildpack apps, [$PORT is exposed][1] as part of slugrunner's job. For Dockerfile/Docker images, the internal port numbers are already known to the developer, so if they want to follow 12-factor best practices then they will have set up $PORT in their own image. I tried seeing if there was a way to work around the docker API's limitations but didn't find one, and then it dawned on me that this is only relevant for users which *do not have control over the exposed ports*. For Dockerfiles/Docker images, this is managed by them so this logic was over-engineered.

deis/publisher exposes the port with the lowest numerical value to the router (docker sorts the internal ports before exposing them in the API), which is why the documentation is reflected to show this.

[1]: https://github.com/deis/deis/blob/e6a078b4119887b4de73e8a6d04291b7cdeb823b/builder/image/slugrunner/Dockerfile#L10-L12

closes #1156. We've been able to run multi-port containers for a while, it's just not what users would expect. This bug was (apparently) gone a long time ago. This is just updating the docs to reflect how to use it in today's world.

Official multi-port support is still untouched: #2095 #3069

### CODE REVIEW

**Comment by bacongobbler** [comment]
Note that I tested and confirmed this by adding `EXPOSE 5001` in my Dockerfile for example-dockerfile-python. Worked great!

File: docs/using_deis/using-docker-images.rst [hunk 1]
```diff
@@ -30,14 +30,10 @@ Docker Image Requirements
 ^^^^^^^^^^^^^^^^^^^^^^^^^
 In order to deploy Docker images, they must conform to the following requirements:
 
- * The Docker image must EXPOSE only one port
- * The exposed port must be an HTTP service that can be connected to an HTTP router
+ * The Docker image must EXPOSE at least one port
```
Comment (bacongobbler) [inline docs/using_deis/using-docker-images.rst]: Not sure how to clarify that official multi-port images (i.e exposing two HTTP processes in a container, or a web and an SSH process once #1315 is tackled) are still unsupported. Any help on this would be appreciated :)

**Comment by carmstrong** [comment]
While this updates the documentation to be more in-line with what's actually happening, it seems to reveal that publisher is doing something that doesn't really make sense. It's not hard to imagine a user asking "Why use the lowest port?" when reading this documentation.

Instead of making this change, perhaps we should reconsider what publisher is doing here? Is supporting multiple ports (#3069) something we should consider doing?

**Comment by bacongobbler** [comment]
It'd be a nice feature for sure! Maybe a happy medium then would to *not* point out how publisher works at the time being (by taking out the note) and consider multi-port apps still unsupported? I was doing a fly-by cleanup more than anything. Multi-port support would be a larger overhaul but I'd be happy to hash out a formal proposal. This just cleans up a lot of the ugly scheduler logic that wasn't necessary here

**Comment by eMxyzptlk** [comment]
Will this allow me to deploy an app that does not expose any ports (think of a data-manipulation micro-service)? Up until now, I had to run a Go server that has no handlers to get out of this problem.

I love that `$PORT` is going to get killed!

**Comment by bacongobbler** [comment]
> Will this allow me to deploy an app that does not expose any ports (think of a data-manipulation micro-service)? Up until now, I had to run a Go server that has no handlers to get out of this problem.

Deis still deploys a `web` process by default, though it's not mandatory. You can always scale it down to 0 and your `worker` process up to 1. This has been the workaround that others have used to deploy apps which don't require exposing any endpoints.

> I love that $PORT is going to get killed!

$PORT's never going to get killed while we're keeping compatibility with Heroku :) There was some overlap between the scheduler and slugrunner (the image that runs Heroku apps) setting up $PORT, so I removed the former for the latter: https://github.com/deis/deis/blob/master/builder/image/slugrunner/Dockerfile#L11-L12

**Comment by carmstrong** [comment]
> Maybe a happy medium then would to not point out how publisher works at the time being (by taking out the note) and consider multi-port apps still unsupported?

:+1:

> This just cleans up a lot of the ugly scheduler logic that wasn't necessary here

Yeah, that cleanup is definitely nice. Is that no longer necessary because publisher does the inspection?

I'm in favor of leaving the documentation as-is and cleaning up the app template, but I'd like to hear what @mboersma thinks.

**Comment by bacongobbler** [comment]
Yes, publisher inspects the ports and exposes the first port it finds (not the best way to solve it, but it works for single-port apps)

**Comment by mboersma** [comment]
I think the behavior of "lowest-numbered port wins" is acceptable for now if it's documented, as it is in this PR. And if I read correctly, the fleet.py unit template code removed here wasn't doing anything anyway (and was very hard to parse). So LGTM IMHO.

**Comment by carmstrong** [comment]
> I think the behavior of "lowest-numbered port wins" is acceptable for now if it's documented, as it is in this PR

The question is *should* this be explicitly documented to be the behavior, or should the language stay that we don't support multiple-port Dockerfiles. I prefer the latter, as that's less confusing for users.

**Comment by mboersma** [comment]
Since there isn't official multi-port support and Deis intends to do that properly at some point, maybe it is counterproductive to document the current behavior. Could we drop that change from the external docs, but add code comments at https://github.com/deis/deis/blob/master/publisher/server/publisher.go#L102 that "lowest-numbered port wins" is the current behavior? tl:dr I agree with @carmstrong now.

**Comment by bacongobbler** [comment]
sure :)

**Comment by eMxyzptlk** [comment]
> Deis still deploys a web process by default, though it's not mandatory. You can always scale it down to 0 and your worker process up to 1. This has been the workaround that others have used to deploy apps which don't require exposing any endpoints.

Interesting, I never knew that. I do not wish to derail the discussion here much but I have one question about this, aren't all of the apps with a Dockerfile tied to one process (the cmd)? So can you actually define multiple Dockerfiles for multiple process types?

> $PORT's never going to get killed while we're keeping compatibility with Heroku :) There was some overlap between the scheduler and slugrunner (the image that runs Heroku apps) setting up $PORT, so I removed the former for the latter: https://github.com/deis/deis/blob/master/builder/image/slugrunner/Dockerfile#L11-L12

I understand, but I meant killed from being in the environment automagically when we clearly have a Dockerfile. It caused me some getting used to that the port being exposed is going to be in $PORT and I can't own that variable.

**Comment by bacongobbler** [comment]
> aren't all of the apps with a Dockerfile tied to one process (the cmd)?

You can supply a Procfile to define multiple process types for an application, even for Dockerfile-based applications: http://docs.deis.io/en/latest/using_deis/process-types/#deploying-to-deis

> So can you actually define multiple Dockerfiles for multiple process types?

All process types share the same container image, but they run in separate containers (again, a Heroku thing).

**Comment by bacongobbler** [comment]
pushed the doc edit as requested

**Comment by carmstrong** [comment]
Code LGTM.

**Comment by mboersma** [comment]
Code LGTM.
